cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java

```diff
@@ -85,76 +87,172 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
   @POST
   @Path("/metadata")
   public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
-    String requestBody = request.content().toString(StandardCharsets.UTF_8);
+    Map<String, GenericRecord> consumeRequests = decodeConsumeRequest(request);
+    MessagesBodyProducer messagesBodyProducer = new MessagesBodyProducer(cConf, consumeRequests, messageFetcher);
```

Comment:
It's probably get overlook in previous review, but the `messageFetcher` should be get from the `MessagingContext` every time when it is needed, instead of sharing the same instance across different thread.